### DISCUSSION TITLE
Partial Derivatives of each DNN output

### DISCUSSION THREAD

**Discussion by dang-trinh156** [lululxvi/deepxde#1248] [Q&A] [answered]
Dear DeepXDE developers and users,

I'm working on solving 
![image](https://user-images.githubusercontent.com/124559501/232290422-3a404770-05fe-4db2-a161-af98b6200aab.png)
![image](https://user-images.githubusercontent.com/124559501/232290448-0d941a0a-8db8-438a-a235-22b403e07535.png)
 this problem, with μo,Bo,μw,Bw is a function of Sw (f(Sw(x,t)) and Kro,Krw is a function of Po (f(Po(x,t)). But as far as I know, deepxde.gradients.jacobian only takes input of all DNN output variables (Po, Sw). How can I compute derivatives of μo,Bo,μw,Bw,Kro,Krw with respect to DNN inputs (x,t) individually?

**Comment by tsarikahin** [comment]
why don't you define them as function of `p_o` and leave the rest to DeepXDE?

For instance:

k_ro = 2*p_o*Z

Assume that your outputs are p_o, Z, and phi (in order), so you have to define your pde:

def pde(x,y)

     k_ro =  2*y[:, 0:1]*y[:,1:2]
     ... 
     
that's all basically.

**Reply by dang-trinh156** [reply]
@tsarikahin I have experimented with DeepXDE example  (Beltrami_flow.py: https://github.com/lululxvi/deepxde/blob/master/examples/pinn_forward/Beltrami_flow.py) about multiplying outputs together: I chang u to u_vel*v_vel in line 14, and it takes forever to train 1000 epoch (I've kept my laptop on hold for 12 hours). I check on Task manager and it doesn't use GPU but CPU instead. But when I differentiate individually, this process utilizes GPU. Here are what I have changed and, CPU not working
![image](https://user-images.githubusercontent.com/124559501/232643539-78d9a9c6-0153-414d-bdb5-1329d531d454.png)
![image](https://user-images.githubusercontent.com/124559501/232643565-e7b43b0f-67e3-4abf-86ce-6e867e0af65a.png)
P/s: I have used the original code of Beltrami_flow.py, it fully utilizes GPU (to 100% during training) and the modification in line 14 just for experimenting. Also, I have checked if Tensorflow is using my GPU and yes, it is using:
![image](https://user-images.githubusercontent.com/124559501/232644296-513c4531-e8d8-4e19-ae3e-b493f5fab94e.png)

**Reply by tsarikahin** [reply]
what about `tensorflow.compat.v1` backend?

**Reply by dang-trinh156** [reply, marked as answer]
@tsarikahin Yeah, thank you for your recommendation! I tried pytorch backend, and it is able to run normally. But one more question, would dde.grad.jacobian  still accept operations on individual cells like torch.where() as normal input ? I mean, would it still be as output1*output2 (u_vel *v_vel in beltrami flow example)?

**Reply by tsarikahin** [reply]
Sure it should be possible as long as the dimensions do match.  You can do column vector multiplication without any problem, e.g. `x[:,0:1]*x[:,1:2]` so the result would be in a shape [:,0:1]. I have done it before.
